[CLEANUP] Various Web Client Platform refactorings

dcleao opened the pull request:
## Various Web Client Platform refactorings
* Naming
  * Moved several implementation classes to `impl` subfolders.
  * Use of new naming scheme for named implementation modules, e.g. `pentaho/config/impl/instanceOfAmdLoadedService`
* Refactored the **context** types
  * Moved to `pentaho/context`
  * Making `pentaho/context` a proper JS class, and a more structured one.
  * Added most other remaining global variables that `webcontext.js` defines to `pentaho.context.spec.IVariables`.
  * Platform context info is configurable through "pentaho/context" AMD configuration
* Promoted the **config** types, from `pentaho/type/config` to `pentaho/config`
  * Global configuration service instance is available at `pentaho/service!pentaho.config.IService?single`
  * Config service no longer converts short type ids to full ids
  * Config service no longer accepts no type selector variable (used to default it to `pentaho/type/value`)
  * `IRuleSelector#type` is now required.
* Refactored the **debug** types.
  * Moved to `pentaho/debug` folder
  * Defined the `pentaho.debug.IManager` interface, which is implemented by the class `pentaho.debug.impl.Manager`,
  * Defined a system `pentaho/debug` singleton module, an instance of `pentaho.debug.impl.Manager`
  * Cannot use `pentaho/service` at this low level, as it would bring problems to AMD loader plugins, within r.js.
* `new pentaho.type.Context(contextVars)` now creates an instance of `pentaho.context.impl.CustomVariables` when receiving a plain object,
   allowing for simpler application container code.
* Various:
  * Commenting misbehaved dojo unit test.
  * Rewrote unit tests clearing the services configuration,
    because now, other unit tests depend on the default services configurations.
  * Rewrote `pentaho/service` unit tests to use the require-test helper
  * Created `pentaho/util/domWindow` and `pentaho/util/global` to be able to mock these when testing
  * Cleanup the test config scripts, after the latest conversion to maven

@pentaho/millenniumfalcon please review.

Merge together with:
* https://github.com/pentaho/pentaho-analyzer/pull/1359
* https://github.com/pentaho/pentaho-platform-plugin-geo/pull/134
* https://github.com/pentaho/pentaho-det/pull/579
* https://github.com/pentaho/eng-sandbox/pull/4
* https://github.com/webdetails/cdf/pull/908

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/_namespace.jsdoc
  doc-jsdoc/pentaho/config/spec/IRule.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSet.jsdoc
  doc-jsdoc/pentaho/type/spec/ITypeProto.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  doc-jsdoc/pentaho/util/_namespace.jsdoc
  doc-jsdoc/pentaho/util/debug/_namespace.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/impl/AmdLoadedService.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/config/impl/serviceInstance.js
  package-res/resources/web/pentaho/config/util.js
  package-res/resources/web/pentaho/type/Context.js
  package-res/resources/web/pentaho/type/_type.js
  package-res/resources/web/pentaho/type/util.js
  package-res/resources/web/pentaho/visual/base/model.js
  test-js/unit/pentaho/config/impl/AmdLoadedService.Spec.js
  test-js/unit/pentaho/config/impl/Service.Spec.js
  test-js/unit/pentaho/config/impl/serviceInstance.Spec.js
  test-js/unit/pentaho/service.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
  test-js/unit/pentaho/visual/base/View.Spec.js
  test-js/unit/require-config.js
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 5.88%
 * Instruction Change:  + 5.33%
 * Line Change:  + 6.67%
 * Method Change:  + 5.88%

nantunes commented on package-res/resources/web/common-ui-require-js-cfg.js:
```diff
@@ -63,6 +64,14 @@
 
+  // Singleton interface, virtual module ids
+  //requireMap["*"]["pentaho/spec/IContext"] = "pentaho/service!pentaho.spec.IContext?single";
+  requireMap["*"]["pentaho/config/IService"] = "pentaho/service!pentaho.config.IService?single";
```
Like I told you, the module ID `pentaho/config/IService` seems strange. I like the map approach (@graimundo doesn't), but only if we could find a better nomenclature for these _default instances_.

dcleao replied:
I also see now why it is confusing...
1. `pentaho/config/IService.instance`
2. `pentaho/config/IService.the`
3. `pentaho/config/IService.first`
4. `pentaho/config/IService.default`
5. `pentaho/config/iService` 
6. `pentaho/config/the.IService`

I'll create a poll for this on slack.

nantunes replied:
We still need to re-address this.

dcleao commented:
Rebased on master to remove from the commit list those of the meanwhile merged base PR.

dcleao commented on doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc:
```diff
@@ -0,0 +1,74 @@
+ * Besides the [type]{@link pentaho.config.spec.IRuleSelector#type} property,
+ * this interface allows filtering on any of the [context properties]{@link pentaho.spec.IContextVars}.
+ *
+ * @name pentaho.config.spec.IRuleSelector
```
Please note "selection" to "selector" change in the rename from `IConfigurationRuleSelection` to `IRuleSelector`.

dcleao commented on package-res/resources/web/common-ui-require-js-cfg.js:
```diff
@@ -63,6 +64,14 @@
+  // Default implementations
+  requireService["pentaho/config/impl/serviceInstance"] = "pentaho.config.IService";
+  requireService["pentaho/config/impl/AmdLoadedService"] = "Class<pentaho.config.IService>";
+
```
Of interest here, and in moved files below, is the experiment of moving auxiliary classes that implement interfaces to a sub-package named `impl`. What do you think?

graimundo replied:
We don't necessarily need to use the same nomenclature but on the server side that is the  guideline that is being used. We include both *api* and *impl* explicitly in the namespace (e.g. _org.pentaho.det.api.query.model.IDataSource_ / _org.pentaho.det.impl.query.model.DataSource_ ).

dcleao replied:
I dislike the use of `api` — the regular namespaces _are the API_. The API side can contain interfaces, mandatory base classes (abstract or not), enums.
A sub-folder named `impl`, anywhere in a regular namespace, would contain auxiliary types, public and/or private, that support implementations of API types.

Ultimately, the impl subfolders could be split into a separate source root. This wouldn't map to a tree without "impl" subfolders, like would happen if "impl" could be implied... Could be used, however, to evidence separation and make it easier to distribute as separate bundles...

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/_namespace.jsdoc
  doc-jsdoc/pentaho/config/spec/IRule.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSet.jsdoc
  doc-jsdoc/pentaho/type/spec/ITypeProto.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  doc-jsdoc/pentaho/util/_namespace.jsdoc
  doc-jsdoc/pentaho/util/debug/_namespace.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/impl/AmdLoadedService.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/config/impl/serviceInstance.js
  package-res/resources/web/pentaho/config/util.js
  package-res/resources/web/pentaho/type/Context.js
  package-res/resources/web/pentaho/type/_type.js
  package-res/resources/web/pentaho/type/util.js
  package-res/resources/web/pentaho/visual/base/model.js
  test-js/unit/pentaho/config/impl/AmdLoadedService.Spec.js
  test-js/unit/pentaho/config/impl/Service.Spec.js
  test-js/unit/pentaho/config/impl/serviceInstance.Spec.js
  test-js/unit/pentaho/service.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
  test-js/unit/pentaho/visual/base/View.Spec.js
  test-js/unit/require-config.js
```

dcleao commented on package-res/resources/web/pentaho/config/util.js:
```diff
@@ -0,0 +1,222 @@
+   * @classDesc A singleton class that contains utilities related to configuration.
+   *
+   * @name pentaho.config.Util
+   * @class
```
This `mergeSpecs` utility method was in the local `pentaho.type.Util` package. I ended creating another local utilities class for the config package: `pentaho.config.Util`. I pondered whether this could be a method of one of the existing utility "pentaho.util.*" but couldn't figure out a good receiver. I pondered creating a new "pentaho/util/spec" utility module, but this would put into question the current rule that "spec is a sub-package of auxiliary, JSON-like interfaces"... Then, this package is also public, and `pentaho.util.*` is almost totally private, the exception being the `MessageBundle` package. There is also no precedent of one of the utility packages in `pentaho.util` having only one public method... Big mess.

nantunes replied:
I think this could easily belong in `pentaho.util`. A potential receiver could be `object.js`, but I would probably prefer it to live by its own (we just need a good name). Other methods could be publicly exposed, like `cloneOwnDeep`.

Lets keep this here for now, but take a note that this should be refactored out either when similar functionality is needed elsewhere or when doing an overall revision for producing separate bundles.

dcleao commented on test-js/unit/pentaho/config/impl/AmdLoadedService.Spec.js:
```diff
@@ -34,19 +34,25 @@ define([], function() {
             {
-              select: {},
+              select: {
+                type: "A"
```
`type` is now mandatory.

nantunes replied:
An empty `type` could still mean _apply to everything_, but I'm also ok with making it mandatory.

dcleao replied:
Before we assumed a default of `pentaho.type.Instance`. To not have a lib specific default, I made it mandatory, but could as well have make it allow `null`. The ConfigService implementation would need to be revised, as I think it indexes rules by type. The empty string could be indexing value. 
I think that the biggest reason is that there seems to be no use case for this.
In a way, it should generally be considered a "bad practice", as we will be injecting properties into types like the JavaScript "Object".
However, if we actually start supporting types and inheritance (from `typeDecls`), "*", "" or "any" would correspond to the JavaScript's "Object" type. Specifying the latter would achieve the desired effect. (All types implicitly inherit from Object, even if not declared).

dcleao commented on package-res/resources/web/pentaho/config/impl/serviceInstance.js:
```diff
@@ -14,26 +14,26 @@
-   * @name pentaho.type.configurationService
-   * @type {pentaho.type.IConfigurationService}
-   * @amd pentaho/type/configurationService
+   * @name pentaho.config.impl.serviceInstance
```
Talvez este package devesse chamar-se `serviceDefault` ?

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/spec/IContextVars.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/iService.js
  package-res/resources/web/pentaho/config/impl/AmdLoadedService.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/contextVars.js
  package-res/resources/web/pentaho/iPlatformContext.js
  package-res/resources/web/pentaho/impl/CustomPlatformContext.js
  package-res/resources/web/pentaho/impl/GlobalPlatformContext.js
  package-res/resources/web/pentaho/type/Context.js
  test-js/unit/pentaho/config/iService.Spec.js
  test-js/unit/pentaho/config/impl/AmdLoadedService.Spec.js
  test-js/unit/pentaho/iPlatformContext.Spec.js
  test-js/unit/pentaho/impl/CustomPlatformContext.Spec.js
  test-js/unit/pentaho/impl/GlobalPlatformContext.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
```

dcleao commented:
1. Rebased.
2. Added `ContextVars#basePath` to eliminate some wild CONTEXT_PATH references still around.
3. Renamed `ContextVars` and related modules to `PlatformContext` — this is to match a corresponding rename of `pentaho/type/Context` to `pentaho/type/Domain` (similar to .Net's Application Domain, or Java's Class Loader) and avoid the confusion that arises from the two contexts actually being two different things.
4. Moved `pentaho/config/util` to `pentaho/util/spec`

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/type/_type.js
  package-res/resources/web/pentaho/util/spec.js
  package-res/resources/web/pentaho/visual/base/model.js
```

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/spec/IContextVars.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/config/impl/instanceOfAmdLoadedService.js
  package-res/resources/web/pentaho/contextVars.js
  package-res/resources/web/pentaho/impl/CustomContextVars.js
  package-res/resources/web/pentaho/impl/GlobalContextVars.js
  package-res/resources/web/pentaho/type/Context.js
  test-js/unit/pentaho/contextVars.Spec.js
  test-js/unit/pentaho/impl/CustomContextVars.Spec.js
  test-js/unit/pentaho/impl/GlobalContextVars.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
```

nantunes commented on package-res/resources/web/common-ui-require-js-cfg.js:
```diff
@@ -26,49 +29,55 @@
       // TODO: This fallback logic is temporary, and can be removed when the remaining
       //    parts of the system rename the "service" plugin id to "pentaho/service".
-      requireService = requireCfg.config["pentaho/service"] || (requireCfg.config["pentaho/service"] = {});
+      requireTypes = requireCfg.config["pentaho/service"] || (requireCfg.config["pentaho/service"] = {});
```
Why the name change?

dcleao replied:
You know :-) Just testing how the conversion to module types looks like.

dcleao commented:
**Summary of Changes**

1. Promoting `pentaho/type/config` to `pentaho/config`
    * Global configuration service instance available at `pentaho/service!pentaho.config.IService?single`
    * Config service no longer converts short ids to full ids
    * Config service no longer accepts no type selector variable (used to default it to pentaho/type/value)
    * Some of the renames:
        * `pentaho.type.IConfigurationService` --> `pentaho.config.IService`
        * `pentaho.type.spec.ITypeConfiguration` --> `pentaho.config.spec.IRuleSet`
        * `pentaho.type.spec.ITypeConfigurationRule` --> `pentaho.config.spec.IRule`
        * `pentaho.type.spec.ITypeConfigurationRuleSelection` --> `pentaho.config.spec.IRuleSelector`
        * `pentaho/util/spec` now contains the objects merge utility (moved code from `pentaho/type/util`)
2. Addition of `pentaho.spec.IContextVars#basePath`
    * Reads its value from the global `CONTEXT_PATH` variable.
3. `requireService = requireCfg.config["pentaho/service"]` --> `requireTypes`
4. Use of `impl` namespace for non-api stuff (e.g. `pentaho/config/impl/instanceOfAmdLoadedService`)
5. new pentaho.type.Context(contextVars) now creates an instance of CustomContextVars, when given a plain object,
   allowing for simpler application code

@pentaho/millenniumfalcon please review

nantunes commented on doc-jsdoc/pentaho/config/IService.jsdoc:
```diff
@@ -15,33 +15,33 @@
  *
- * The [select]{@link pentaho.type.IConfigurationService#select} method
- * selects the rules of added type configurations that apply to
+ * The [select]{@link pentaho.config.IService#select} method
```
I still prefer _extracting_ more information from the class name itself, without knowing the package. I mean `IConfigurationService`. But is just a matter of personal taste, not a blocker.

dcleao commented:
@nantunes Yes, some changes are still needed in other projects, like DET and Analyzer. I'll handle those as soon as we stabilize on these changes.

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/spec/IContextVars.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/config/impl/instanceOfAmdLoadedService.js
  package-res/resources/web/pentaho/contextVars.js
  package-res/resources/web/pentaho/impl/CustomContextVars.js
  package-res/resources/web/pentaho/impl/GlobalContextVars.js
  package-res/resources/web/pentaho/type/Context.js
  test-js/unit/pentaho/contextVars.Spec.js
  test-js/unit/pentaho/impl/CustomContextVars.Spec.js
  test-js/unit/pentaho/impl/GlobalContextVars.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
```

Newly Fixed Tests:
======================
###### org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson1:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: wrong value
	at org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson1(MetadataService2IT.java:731)

```
###### org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson2:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: wrong value
	at org.pentaho.common.ui.metadata.service.MetadataService2IT.testJsonQueryToJson2(MetadataService2IT.java:750)

```
###### org.pentaho.common.ui.metadata.service.MetadataService2IT.testXmlQueryToJson1:

:large_blue_circle: `junit.framework.AssertionFailedError`

```java
junit.framework.AssertionFailedError: wrong value
	at org.pentaho.common.ui.metadata.service.MetadataService2IT.testXmlQueryToJson1(MetadataService2IT.java:711)

```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 11.76%
 * Instruction Change:  + 10.67%
 * Line Change:  + 13.33%
 * Method Change:  + 11.76%

wingman-pentaho commented:
### Build Completed

:fire: This pull request has some issues. It would be preferable to fix them in order for it to be just perfect. See below for more details. Some links are also available below for further assistance in addressing those issues.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  test-js/unit/dojo/pentaho/DojoNumberSpec.js
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change: **-11.76%**:small_red_triangle_down:
 * Instruction Change: **-10.67%**:small_red_triangle_down:
 * Line Change: **-13.33%**:small_red_triangle_down:
 * Method Change: **-11.76%**:small_red_triangle_down:

dcleao commented:
@pentaho/millenniumfalcon I rebased on master and squashed the commits — it was becoming a bit unwieldy. I apologize to those that already reviewed previous commits.
I rewrote the text on the now single commit to summarize all of the changes and also updated the PR description.
Note again that this requires some changes in other projects, which I will prepare as soon as we settle here.  Thanks!

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  config/context-begin.js
  doc-jsdoc/pentaho/IDebugManager.jsdoc
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/_namespace.jsdoc
  doc-jsdoc/pentaho/config/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/config/spec/IRule.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSet.jsdoc
  doc-jsdoc/pentaho/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/spec/IContextVars.jsdoc
  doc-jsdoc/pentaho/type/spec/ITypeProto.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  doc-jsdoc/pentaho/util/_namespace.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/DebugLevels.js
  package-res/resources/web/pentaho/config/impl/AmdLoadedService.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/config/impl/instanceOfAmdLoadedService.js
  package-res/resources/web/pentaho/contextVars.js
  package-res/resources/web/pentaho/debug.js
  package-res/resources/web/pentaho/impl/CustomContextVars.js
  package-res/resources/web/pentaho/impl/DebugManager.js
  package-res/resources/web/pentaho/lang/Base.js
  package-res/resources/web/pentaho/type/Context.js
  package-res/resources/web/pentaho/type/_type.js
  package-res/resources/web/pentaho/type/util.js
  package-res/resources/web/pentaho/util/domWindow.js
  package-res/resources/web/pentaho/util/spec.js
  package-res/resources/web/pentaho/visual/base/model.js
  test-js/unit/dojo/pentaho/DojoNumberSpec.js
  test-js/unit/pentaho/DebugLevels.Spec.js
  test-js/unit/pentaho/config/impl/AmdLoadedService.Spec.js
  test-js/unit/pentaho/config/impl/Service.Spec.js
  test-js/unit/pentaho/contextVars.Spec.js
  test-js/unit/pentaho/debug.Spec.js
  test-js/unit/pentaho/impl/CustomContextVars.Spec.js
  test-js/unit/pentaho/impl/DebugManager.Spec.js
  test-js/unit/pentaho/service.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
  test-js/unit/pentaho/visual/base/View.Spec.js
  test-js/unit/require-config.js
  test-js/unit/test-utils.js
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 5.88%
 * Instruction Change:  + 5.33%
 * Line Change:  + 6.67%
 * Method Change:  + 5.88%

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/context/_namespace.jsdoc
  doc-jsdoc/pentaho/context/spec/IVariables.jsdoc
  doc-jsdoc/pentaho/context/spec/_namespace.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/context/impl/CustomVariables.js
  package-res/resources/web/pentaho/context/vars.js
  package-res/resources/web/pentaho/i18n.js
  package-res/resources/web/pentaho/type/Context.js
  test-js/unit/pentaho/context/impl/CustomVariables.Spec.js
  test-js/unit/pentaho/context/vars.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
```

Newly Fixed Tests:
======================
###### org.pentaho.common.ui.metadata.service.MetadataServiceIT.testCondition2:

:large_blue_circle: `flexjson.JSONException`

```java
flexjson.JSONException: Error trying to deepSerialize
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransformer.java:61)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.JSONContext.transform(JSONContext.java:73)
	at flexjson.transformer.ArrayTransformer.transform(ArrayTransformer.java:31)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransform
```
###### org.pentaho.common.ui.metadata.service.MetadataServiceIT.testGetModelJson:

:large_blue_circle: `flexjson.JSONException`

```java
flexjson.JSONException: Error trying to deepSerialize
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransformer.java:61)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.JSONContext.transform(JSONContext.java:73)
	at flexjson.transformer.ArrayTransformer.transform(ArrayTransformer.java:31)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransform
```
###### org.pentaho.common.ui.metadata.service.MetadataServiceIT.testJsonQuery:

:large_blue_circle: `flexjson.JSONException`

```java
flexjson.JSONException: Error trying to deepSerialize
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransformer.java:61)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.JSONContext.transform(JSONContext.java:73)
	at flexjson.transformer.ArrayTransformer.transform(ArrayTransformer.java:31)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransform
```
###### org.pentaho.common.ui.metadata.service.MetadataServiceIT.testJsonQueryToCdaJson1:

:large_blue_circle: `flexjson.JSONException`

```java
flexjson.JSONException: Error trying to deepSerialize
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransformer.java:61)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.JSONContext.transform(JSONContext.java:73)
	at flexjson.transformer.ArrayTransformer.transform(ArrayTransformer.java:31)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransform
```
###### org.pentaho.common.ui.metadata.service.MetadataServiceIT.testJsonQueryToJson1:

:large_blue_circle: `flexjson.JSONException`

```java
flexjson.JSONException: Error trying to deepSerialize
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransformer.java:61)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.JSONContext.transform(JSONContext.java:73)
	at flexjson.transformer.ArrayTransformer.transform(ArrayTransformer.java:31)
	at flexjson.transformer.TransformerWrapper.transform(TransformerWrapper.java:22)
	at flexjson.transformer.ObjectTransformer.transform(ObjectTransform
```

#### Integration test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Column
 * Complexity Change:  + 8.00%
 * Instruction Change:  + 6.12%
 * Line Change:  + 5.26%
 * Method Change:  + 8.00%

###### org.pentaho.common.ui.metadata.model.impl.Condition
 * Branch Change:  + 3.85%
 * Complexity Change:  + 12.90%
 * Instruction Change:  + 3.80%
 * Line Change:  + 6.67%
 * Method Change:  + 16.67%

###### org.pentaho.common.ui.metadata.model.impl.Order
 * Complexity Change:  + 14.29%
 * Instruction Change:  + 12.50%
 * Line Change:  + 10.00%
 * Method Change:  + 14.29%

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 5.88%
 * Instruction Change:  + 4.00%
 * Line Change:  + 3.33%
 * Method Change:  + 5.88%

###### org.pentaho.common.ui.metadata.model.Operator
 * Branch Change:  + 43.75%
 * Complexity Change:  + 60.47%
 * Instruction Change:  + 31.86%
 * Line Change:  + 41.30%

###### org.pentaho.common.ui.metadata.service.MetadataService
 * Complexity Change:  + 9.76%
 * Instruction Change:  + 9.77%
 * Line Change:  + 8.09%
 * Method Change:  + 20.00%

###### org.pentaho.common.ui.metadata.service.MetadataServiceUtil
 * Complexity Change:  + 2.04%
 * Instruction Change:  + 2.59%
 * Line Change:  + 2.50%
 * Method Change:  + 7.14%

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  config/context-begin.js
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/_namespace.jsdoc
  doc-jsdoc/pentaho/config/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/config/spec/IRule.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSet.jsdoc
  doc-jsdoc/pentaho/context/_namespace.jsdoc
  doc-jsdoc/pentaho/context/spec/IVariables.jsdoc
  doc-jsdoc/pentaho/context/spec/_namespace.jsdoc
  doc-jsdoc/pentaho/debug/IManager.jsdoc
  doc-jsdoc/pentaho/debug/_namespace.jsdoc
  doc-jsdoc/pentaho/debug/impl/_namespace.jsdoc
  doc-jsdoc/pentaho/type/spec/ITypeProto.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextVarFilter.jsdoc
  doc-jsdoc/pentaho/util/_namespace.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/impl/AmdLoadedService.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/config/impl/instanceOfAmdLoadedService.js
  package-res/resources/web/pentaho/context/impl/CustomVariables.js
  package-res/resources/web/pentaho/context/vars.js
  package-res/resources/web/pentaho/debug/Levels.js
  package-res/resources/web/pentaho/debug/impl/Manager.js
  package-res/resources/web/pentaho/debug/manager.js
  package-res/resources/web/pentaho/i18n.js
  package-res/resources/web/pentaho/lang/Base.js
  package-res/resources/web/pentaho/type/Context.js
  package-res/resources/web/pentaho/type/_type.js
  package-res/resources/web/pentaho/type/util.js
  package-res/resources/web/pentaho/util/domWindow.js
  package-res/resources/web/pentaho/util/spec.js
  package-res/resources/web/pentaho/visual/base/model.js
  test-js/unit/dojo/pentaho/DojoNumberSpec.js
  test-js/unit/pentaho/config/impl/AmdLoadedService.Spec.js
  test-js/unit/pentaho/config/impl/Service.Spec.js
  test-js/unit/pentaho/context/impl/CustomVariables.Spec.js
  test-js/unit/pentaho/context/vars.Spec.js
  test-js/unit/pentaho/debug/Levels.Spec.js
  test-js/unit/pentaho/debug/impl/Manager.Spec.js
  test-js/unit/pentaho/debug/manager.Spec.js
  test-js/unit/pentaho/service.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
  test-js/unit/pentaho/visual/base/View.Spec.js
  test-js/unit/require-config.js
  test-js/unit/test-utils.js
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 11.76%
 * Instruction Change:  + 10.67%
 * Line Change:  + 13.33%
 * Method Change:  + 11.76%

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  package-res/resources/web/pentaho/context/vars.js
  package-res/resources/web/pentaho/util/domWindow.js
  package-res/resources/web/pentaho/util/global.js
  test-js/unit/pentaho/context/vars.Spec.js
```

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -Dtestreports.dir=bin/reports/integration-test -f build.xml resolve resolve-js dist jacoco-integration test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/debug/impl/Manager.js
  package-res/resources/web/pentaho/debug/manager.js
  test-js/unit/pentaho/debug/manager.Spec.js
```

nantunes commented on package-res/resources/web/pentaho/util/global.js:
```diff
@@ -0,0 +1,29 @@
+  return (typeof self !== "undefined") ? self :
+         (typeof window !== "undefined") ? window :
+         (typeof global !== "undefined") ? global :
+         Function("return this")();
```
I'm confused on what `Function(...)` does... Shouldn't it be `new Function(...)`.
I know it must be right, I simply don't understand it. 🤔

dcleao replied:
From [MDN, on Function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function):
"Invoking the Function constructor as a function (without using the new operator) has the same effect as invoking it as a constructor."

nantunes replied:
:-D _"Hey, let's just make two of each!"
👍

graimundo commented on package-res/resources/web/pentaho/util/spec.js:
```diff
@@ -0,0 +1,222 @@
+   *
+   * @name pentaho.util.Spec
+   * @class
+   * @amd {pentaho.util.Spec} pentaho/util/spec
```
If they're utilities that pertain configuration shouldn't that be reflected in the namespace somehow? pentaho.util.configuration.Spec or pentaho.config.util.Spec or pentaho.config.Util or another one that refers the configuration? 
Personally I have preference for having it in the configuration namespace => pentaho.config.Util

dcleao replied:
It's a general utility on how to merge **specification** (hence the _Spec_ name), that is used by the configuration modules but also elsewhere on pentaho/type.

graimundo replied:
I should've take a closer look at the code instead of just looking at the class description. 
Given that info I would leave the util where it is but remove any comments that refer configuration.

graimundo commented on test-js/unit/dojo/pentaho/DojoNumberSpec.js:
```diff
@@ -22,10 +22,11 @@ define( ["dojo/number"], function( DojoNumber ) {
 
-    it( "Dojo Number.parse(..) should accept correct locale", function() {
+    // TODO: misbehaved test; does not always work
+    xit( "Dojo Number.parse(..) should accept correct locale", function() {
```
@davidmsantos90  was this one of the tests we were having issues with during the mavenization?

davidmsantos90 replied:
@graimundo only noticed this ping now, but yes, this is the same test

dcleao commented on doc-jsdoc/pentaho/context/spec/IVariables.jsdoc:
```diff
@@ -0,0 +1,119 @@
+/**
+ * The identifier of the client application.
+ *
+ * This is the value specified in a web context script URL, as in: `webcontext.js?context=application-id`.
```
There is one mismatch, that I had noticed but forgot, in the meantime, which is that the Analyzer context passed to webcontext.js is "analyzer" and the application id used by Analyzer to instantiate the `pentaho/type/Context` is `pentaho/analyzer` (or something like that)...
The idea is that there is one main/default context. This context is typically an application that holds a frame. Other non-default contexts may be created as needed (they are just objects that satisfy the `pentaho.context.spec.IVariables` interface).

dcleao replied:
I've come to agree with @graimundo on keeping the two concepts separate. The main/default context will never have an application defined. Applications must create a context where they override its `application` id property. They currently already do this when creating their corresponding `pentaho/type/Context`.

dcleao commented on doc-jsdoc/pentaho/context/spec/IVariables.jsdoc:
```diff
@@ -0,0 +1,119 @@
+ * The home folder of the user.
+ *
+ * @name userHome
+ * @memberOf pentaho.context.spec.IVariables#
```
I would also prefer the grouping you propose on user, reservedChars and server. The reason I did not do that was because our current configuration system does not support "path" variables on selectors.  Also, note on your URL proposal, the toString method would imply an "API" (with methods and such) and the interface is by design as simple as possible, so that it can be provided inline by webcontext.js without having to include classes, before, to do so.

dcleao replied:
Ended up accepting the superior object interface. The only variable currently being used for configurations is `application`, and that name will remain unchanged. Using the `user.id` variable in the future will require support from the configuration service for selecting on "path properties".

dcleao commented:
@pentaho/millenniumfalcon added commit addressing @graimundo comments. Please review.

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
ant -Dtestreports.dir=bin/reports/unit-test -f build.xml clean-all resolve resolve-js jacoco && ant -f build.xml resolve resolve-js dist test-js checkstyle
```

#### Cleanup Commands
```sh
rm -r ~/.ivy2/local || echo no publish local to remove
```

#### Changed files

```console
  doc-jsdoc/pentaho/config/IService.jsdoc
  doc-jsdoc/pentaho/config/spec/IRuleSelector.jsdoc
  doc-jsdoc/pentaho/context/IContext.jsdoc
  doc-jsdoc/pentaho/context/IReservedChars.jsdoc
  doc-jsdoc/pentaho/context/IServer.jsdoc
  doc-jsdoc/pentaho/context/IUser.jsdoc
  doc-jsdoc/pentaho/context/spec/IContext.jsdoc
  doc-jsdoc/pentaho/context/spec/IReservedChars.jsdoc
  doc-jsdoc/pentaho/context/spec/IServer.jsdoc
  doc-jsdoc/pentaho/context/spec/IUser.jsdoc
  doc-jsdoc/pentaho/type/spec/UContextPropFilter.jsdoc
  package-res/resources/web/common-ui-require-js-cfg.js
  package-res/resources/web/pentaho/config/impl/Service.js
  package-res/resources/web/pentaho/context.js
  package-res/resources/web/pentaho/context/impl/Context.js
  package-res/resources/web/pentaho/debug.js
  package-res/resources/web/pentaho/i18n.js
  package-res/resources/web/pentaho/type/Context.js
  package-res/resources/web/pentaho/util/spec.js
  test-js/unit/pentaho/context.Spec.js
  test-js/unit/pentaho/context/impl/Context.Spec.js
  test-js/unit/pentaho/debug.Spec.js
  test-js/unit/pentaho/type/Context.Spec.js
```

#### Unit test coverage change

These statistics help you identify how your changes have affected the coverage of the following files. If a file is not in this list, then its coverage was not affected by your changes. To get some help interpreting these metrics, please refer to [Jacoco's documentation](http://eclemma.org/jacoco/trunk/doc/counters.html).

###### org.pentaho.common.ui.metadata.model.impl.Query
 * Complexity Change:  + 5.88%
 * Instruction Change:  + 5.33%
 * Line Change:  + 6.67%
 * Method Change:  + 5.88%

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
mvn -B -fn -f impl/client/pom.xml clean install && mvn -B -f impl/client/pom.xml site
mvn -B -fn -f assemblies/platform-plugin/pom.xml clean install && mvn -B -f assemblies/platform-plugin/pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f impl/client/pom.xml build-helper:remove-project-artifact && mvn -B -f assemblies/platform-plugin/pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  assemblies/platform-plugin/src/main/resources/resources/web/common-ui-require-js-cfg.js
  impl/client/src/main/doc-js/pentaho/config/IService.jsdoc
  impl/client/src/main/doc-js/pentaho/config/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/config/impl/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRule.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRuleSelector.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRuleSet.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IContext.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IReservedChars.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IServer.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IUser.jsdoc
  impl/client/src/main/doc-js/pentaho/context/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IContext.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IReservedChars.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IServer.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IUser.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/IManager.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/impl/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/type/spec/ITypeProto.jsdoc
  impl/client/src/main/doc-js/pentaho/type/spec/UContextPropFilter.jsdoc
  impl/client/src/main/doc-js/pentaho/util/_namespace.jsdoc
  impl/client/src/main/javascript/web/pentaho/config/impl/AmdLoadedService.js
  impl/client/src/main/javascript/web/pentaho/config/impl/Service.js
  impl/client/src/main/javascript/web/pentaho/config/impl/instanceOfAmdLoadedService.js
  impl/client/src/main/javascript/web/pentaho/context.js
  impl/client/src/main/javascript/web/pentaho/context/impl/Context.js
  impl/client/src/main/javascript/web/pentaho/debug.js
  impl/client/src/main/javascript/web/pentaho/debug/Levels.js
  impl/client/src/main/javascript/web/pentaho/debug/impl/Manager.js
  impl/client/src/main/javascript/web/pentaho/i18n.js
  impl/client/src/main/javascript/web/pentaho/lang/Base.js
  impl/client/src/main/javascript/web/pentaho/type/Context.js
  impl/client/src/main/javascript/web/pentaho/type/_type.js
  impl/client/src/main/javascript/web/pentaho/type/util.js
  impl/client/src/main/javascript/web/pentaho/util/domWindow.js
  impl/client/src/main/javascript/web/pentaho/util/global.js
  impl/client/src/main/javascript/web/pentaho/util/spec.js
  impl/client/src/main/javascript/web/pentaho/visual/base/model.js
  impl/client/src/test/config/javascript/context-begin.js
  impl/client/src/test/config/javascript/context-end.js
  impl/client/src/test/config/javascript/require.config.js
  impl/client/src/test/javascript/pentaho/config/impl/AmdLoadedService.spec.js
  impl/client/src/test/javascript/pentaho/config/impl/Service.spec.js
  impl/client/src/test/javascript/pentaho/context.spec.js
  impl/client/src/test/javascript/pentaho/context/impl/Context.spec.js
  impl/client/src/test/javascript/pentaho/debug.spec.js
  impl/client/src/test/javascript/pentaho/debug/Levels.spec.js
  impl/client/src/test/javascript/pentaho/debug/impl/Manager.spec.js
  impl/client/src/test/javascript/pentaho/service.spec.js
  impl/client/src/test/javascript/pentaho/type/Context.spec.js
  impl/client/src/test/javascript/pentaho/visual/base/View.spec.js
```

dcleao commented:
Rebased on master.

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
mvn -B -fn -f impl/client/pom.xml clean install && mvn -B -f impl/client/pom.xml site
mvn -B -fn -f assemblies/platform-plugin/pom.xml clean install && mvn -B -f assemblies/platform-plugin/pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f impl/client/pom.xml build-helper:remove-project-artifact && mvn -B -f assemblies/platform-plugin/pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  assemblies/platform-plugin/src/main/resources/resources/web/common-ui-require-js-cfg.js
  impl/client/src/main/doc-js/pentaho/config/IService.jsdoc
  impl/client/src/main/doc-js/pentaho/config/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/config/impl/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRule.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRuleSelector.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRuleSet.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IContext.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IServer.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IUser.jsdoc
  impl/client/src/main/doc-js/pentaho/context/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IContext.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IServer.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IUser.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/IManager.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/impl/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/type/spec/ITypeProto.jsdoc
  impl/client/src/main/doc-js/pentaho/type/spec/UContextPropFilter.jsdoc
  impl/client/src/main/doc-js/pentaho/util/_namespace.jsdoc
  impl/client/src/main/javascript/web/pentaho/config/impl/AmdLoadedService.js
  impl/client/src/main/javascript/web/pentaho/config/impl/Service.js
  impl/client/src/main/javascript/web/pentaho/config/impl/instanceOfAmdLoadedService.js
  impl/client/src/main/javascript/web/pentaho/context.js
  impl/client/src/main/javascript/web/pentaho/context/impl/Context.js
  impl/client/src/main/javascript/web/pentaho/debug.js
  impl/client/src/main/javascript/web/pentaho/debug/Levels.js
  impl/client/src/main/javascript/web/pentaho/debug/impl/Manager.js
  impl/client/src/main/javascript/web/pentaho/i18n.js
  impl/client/src/main/javascript/web/pentaho/lang/Base.js
  impl/client/src/main/javascript/web/pentaho/type/Context.js
  impl/client/src/main/javascript/web/pentaho/type/_type.js
  impl/client/src/main/javascript/web/pentaho/type/util.js
  impl/client/src/main/javascript/web/pentaho/util/domWindow.js
  impl/client/src/main/javascript/web/pentaho/util/global.js
  impl/client/src/main/javascript/web/pentaho/util/spec.js
  impl/client/src/main/javascript/web/pentaho/visual/base/model.js
  impl/client/src/test/config/javascript/context-begin.js
  impl/client/src/test/config/javascript/context-end.js
  impl/client/src/test/config/javascript/karma.ci.conf.js
  impl/client/src/test/config/javascript/karma.conf.js
  impl/client/src/test/config/javascript/require-test.js
  impl/client/src/test/config/javascript/require.config.js
  impl/client/src/test/javascript/angular-directives/accordionWizard.spec.js
  impl/client/src/test/javascript/angular-directives/dateTimePicker.spec.js
  impl/client/src/test/javascript/angular-directives/schedule.spec.js
  impl/client/src/test/javascript/angular-directives/weeklyRecurrence.spec.js
  impl/client/src/test/javascript/pentaho/config/impl/AmdLoadedService.spec.js
  impl/client/src/test/javascript/pentaho/config/impl/Service.spec.js
  impl/client/src/test/javascript/pentaho/context.spec.js
  impl/client/src/test/javascript/pentaho/context/impl/Context.spec.js
  impl/client/src/test/javascript/pentaho/debug.spec.js
  impl/client/src/test/javascript/pentaho/debug/Levels.spec.js
  impl/client/src/test/javascript/pentaho/debug/impl/Manager.spec.js
  impl/client/src/test/javascript/pentaho/service.spec.js
  impl/client/src/test/javascript/pentaho/type/Context.spec.js
  impl/client/src/test/javascript/pentaho/visual/base/View.spec.js
```

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/config/spec/IRuleSelector.jsdoc:
```diff
@@ -0,0 +1,74 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
Missed this license header date

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/context/IContext.jsdoc:
```diff
@@ -0,0 +1,119 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/context/IServer.jsdoc:
```diff
@@ -0,0 +1,38 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/context/IUser.jsdoc:
```diff
@@ -0,0 +1,42 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/context/spec/IServer.jsdoc:
```diff
@@ -0,0 +1,31 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/context/spec/IUser.jsdoc:
```diff
@@ -0,0 +1,39 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/debug/IManager.jsdoc:
```diff
@@ -0,0 +1,102 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/debug/_namespace.jsdoc:
```diff
@@ -0,0 +1,22 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/doc-js/pentaho/debug/impl/_namespace.jsdoc:
```diff
@@ -0,0 +1,23 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/javascript/web/pentaho/context/impl/Context.js:
```diff
@@ -0,0 +1,126 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/javascript/web/pentaho/debug.js:
```diff
@@ -0,0 +1,82 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/javascript/web/pentaho/debug/Levels.js:
```diff
@@ -0,0 +1,112 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/javascript/web/pentaho/debug/impl/Manager.js:
```diff
@@ -0,0 +1,82 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

davidmsantos90 commented on impl/client/src/main/javascript/web/pentaho/util/spec.js:
```diff
@@ -0,0 +1,222 @@
+/*!
+ * Copyright 2010 - 2016 Pentaho Corporation. All rights reserved.
```
header date update

wingman-pentaho commented:
### Build Completed

:white_check_mark: This pull request has passed all validations.

#### Build Commands

```sh
mvn -B -fn -f impl/client/pom.xml clean install && mvn -B -f impl/client/pom.xml site
```

#### Cleanup Commands
```sh
mvn -B -f impl/client/pom.xml build-helper:remove-project-artifact
```

#### Changed files

```console
  impl/client/src/main/doc-js/pentaho/config/IService.jsdoc
  impl/client/src/main/doc-js/pentaho/config/impl/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRule.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRuleSelector.jsdoc
  impl/client/src/main/doc-js/pentaho/config/spec/IRuleSet.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IContext.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IServer.jsdoc
  impl/client/src/main/doc-js/pentaho/context/IUser.jsdoc
  impl/client/src/main/doc-js/pentaho/context/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IContext.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IServer.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/IUser.jsdoc
  impl/client/src/main/doc-js/pentaho/context/spec/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/IManager.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/_namespace.jsdoc
  impl/client/src/main/doc-js/pentaho/debug/impl/_namespace.jsdoc
  impl/client/src/main/javascript/web/pentaho/config/impl/AmdLoadedService.js
  impl/client/src/main/javascript/web/pentaho/config/impl/Service.js
  impl/client/src/main/javascript/web/pentaho/config/impl/instanceOfAmdLoadedService.js
  impl/client/src/main/javascript/web/pentaho/context.js
  impl/client/src/main/javascript/web/pentaho/context/impl/Context.js
  impl/client/src/main/javascript/web/pentaho/debug.js
  impl/client/src/main/javascript/web/pentaho/debug/Levels.js
  impl/client/src/main/javascript/web/pentaho/debug/impl/Manager.js
  impl/client/src/main/javascript/web/pentaho/service.js
  impl/client/src/test/javascript/pentaho/config/impl/AmdLoadedService.spec.js
  impl/client/src/test/javascript/pentaho/config/impl/Service.spec.js
  impl/client/src/test/javascript/pentaho/context.spec.js
  impl/client/src/test/javascript/pentaho/context/impl/Context.spec.js
  impl/client/src/test/javascript/pentaho/debug.spec.js
  impl/client/src/test/javascript/pentaho/service.spec.js
```
